sport-moment-nft/contracts/EPL.cdc

```diff
@@ -0,0 +1,1124 @@
+        pub let name: String
+        pub let locked: Bool
+
+        /// initializer
+        ///
+        init (id: UInt64) {
+            let set = (&EPL.setByID[id] as! &EPL.Set?)!
+            self.id = id
+            self.name = set.name
+            self.locked = set.locked
+        }
+    }
+
+    /// A top level Set with a unique ID and a name
+    ///
+    pub resource Set {
+        pub let id: UInt64
+        pub let name: String
+
+        // Indicates if the Set is currently locked.
+        // When a Set is created, it is unlocked
+        // and Editions can be created with it.
+        // When a Set is locked, new Editions cannot be created with the Set.
+        // A Set can never be changed from locked to unlocked,
+        // the decision to lock a Set is final.
+        // If a Set is locked, Moments can still be minted from the
+        // Editions already created from the Set.
+        pub var locked: Bool
+
+        /// initializer
+        ///
+        init (name: String) {
+            pre {
+                !EPL.setIDByName.containsKey(name): "A Set with that name already exists"
+            }
+            self.id = EPL.nextSetID
+            self.name = name
+            self.locked = false
+
+            // Cache the new set's name => ID
+            EPL.setIDByName[name] = self.id
+            // Increment for the nextSeriesID
+            EPL.nextSetID = self.id + 1 as UInt64
+
+            emit SetCreated(id: self.id, name: self.name)
+        }
+
+        // lock() locks the Set so that no more Plays can be added to it
+        //
+        // Pre-Conditions:
+        // The Set should not be locked
+        pub fun lock() {
+            if !self.locked {
+                self.locked = true
+                emit SetLocked(setID: self.id)
+            }
+        }
+    }
+
+    /// Get the publicly available data for a Set
+    ///
+    pub fun getSetData(id: UInt64): EPL.SetData? {
+        if EPL.setByID[id] == nil {
+            return nil
+        }
+        return EPL.SetData(id: id!)
+    }
+
+    /// Get the publicly available data for a Set by name
+    ///
+    pub fun getSetDataByName(name: String): EPL.SetData? {
+        let id = EPL.setIDByName[name]
+
+        if id == nil {
+            return nil
+        }
+        return EPL.SetData(id: id!)
+    }
+
+    /// Get all set names (this will be *long*)
+    ///
+    pub fun getAllSetNames(): [String] {
+        return EPL.setIDByName.keys
+    }
+
+    //------------------------------------------------------------
+    // Tag
+    //------------------------------------------------------------
+    pub struct TagData {
+        pub let id: UInt64
+        pub let name: String
+
+        /// initializer
+        ///
+        init (id: UInt64) {
+            let tag = (&EPL.tagByID[id] as! &EPL.Tag?)!
+            self.id = id
+            self.name = tag.name
+        }
+    }
+
+    /// A top level Tag with a unique ID
+    //
+    pub resource Tag {
```

Comment:
Can you add more comments about what a Tag is?

Reply:
A tag is the type of play or its classification. For soccer a play can contain an assist, a header, a goal, a game winning goal. And it can contain all four of these at once. Tags will classify these types and apply to the play as a list.

Reply:
Done